src/interface/guide/components/CastEfficiencyRibbon.tsx

```diff
@@ -0,0 +1,451 @@
+import styled from '@emotion/styled';
+import { SpellLink, Tooltip } from 'interface';
+import { useAnalyzer, useInfo, useEvents } from 'interface/guide';
+import CastEfficiency from 'parser/shared/modules/CastEfficiency';
+import Abilities from 'parser/core/modules/Abilities';
+import EventHistory from 'parser/shared/modules/EventHistory';
+import { formatPercentage, formatDuration } from 'common/format';
+import Spell from 'common/SPELLS/Spell';
+import { StatCard, StatValue, StatLabel, HelperText } from './GuideDivs';
+import GuideDataWrapper from './GuideDataWrapper';
+import { EventType, UpdateSpellUsableEvent, UpdateSpellUsableType } from 'parser/core/Events';
+import BulletGraph, { PerformanceRange } from './BulletGraph';
+import { CooldownWindow } from 'parser/ui/CooldownBar';
+import SegmentedTimeline, { TimelineSegment, TimelineMarker } from './SegmentedTimeline';
+import { BAD_COLOR, getSpecColor } from 'interface/guide/colors';
+
+// Styled Components
+const RibbonContainer = styled.div`
+  background: rgba(0, 0, 0, 0.3);
+  border-radius: 6px;
+  padding: 8px 12px;
+  margin-top: 8px;
+`;
+
+const TimelineWrapper = styled.div`
+  width: 100%;
+  height: 40px;
+  display: flex;
+  align-items: center;
+`;
+
+// Helper functions to convert combat log events to timeline segments/markers
+function createCooldownSegments(
+  spellId: number,
+  events: ReturnType<typeof useEvents>,
```

Comment:
nit: this is clearer as `AnyEvent[]`
```suggestion
  events: AnyEvent[],
```